### PR TITLE
Add contributing dir with Config file checklist

### PR DESCRIPTION
**Pull request by banks**
This checklist was reviewed internally in a document form a few months ago but I wanted to make it public.

I've followed Nomad's conventions of adding a `contributing` dir. The only quirk here is we already have a `.github/CONTRIBUTING.md` which has some overlap. GitHub links to that in new PRs so it's useful to surface there and will only pick it up from `.github` in the root or in `/docs/` dir. I don't want to add `/docs` and choose not to move it for now as we don't want other checklists in the root so it still doesn't end up in the same place but open to other options here.

I tried to make this a relatively minimal change and we can certainly iterate on the README and other things.

The `contributing/README.md` is based heavily on the Nomad one: https://github.com/hashicorp/nomad/tree/master/contributing for now. I don't know that they need to be in perfect harmony so we can iterate but it seemed like a decent starting point.

Finally I moved `INTERNALS.md` to avoid having a sprawl of things in different places I left a stub behind so links don't break in case anyone links to it in mailing list issues etc.

### CODE REVIEW

File: contributing/checklist-adding-config-fields.md [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+ - [ ] Add the new field with a random value to both the JSON and HCL blobs in
+   `TestFullConfig` in `agent/config/runtime_test.go`, it should fail now, then
+   add the same random value to the expected struct in that test so it passes
+   again.
+ - [ ] Add the new field and it's default value to `TestSanitize` in the same
+   file. (Running the test first gives you a nice diff which can save working
+   out where etc.)
+ - [ ] **If** your new config field needed some validation as it's only valid in
+   some cases or with some values (often true).
+      - [ ] Add validation to Validate in `builder.go`.
```
Comment (crhino) [inline contributing/checklist-adding-config-fields.md]:
```suggestion
      - [ ] Add validation to Validate in `agent/config/builder.go`.
```

File: contributing/checklist-adding-config-fields.md [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+      - [ ] Add a test case to the table test `TestConfigFlagsAndEdgeCases` in
+        `runtime_test.go`.
```
Comment (crhino) [inline contributing/checklist-adding-config-fields.md]:
```suggestion
        `agent/config/runtime_test.go`.
```

File: contributing/checklist-adding-config-fields.md [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+ - [ ] **If** your new config field needs a non-zero-value default.
+      - [ ] Add that to `DefaultSource` in `defaults.go`.
```
Comment (crhino) [inline contributing/checklist-adding-config-fields.md]:
```suggestion
      - [ ] Add that to `DefaultSource` in `agent/config/defaults.go`.
```

File: contributing/checklist-adding-config-fields.md [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+      - [ ] Add a test case to the table test `TestConfigFlagsAndEdgeCases` in
+        `runtime_test.go`.
```
Comment (crhino) [inline contributing/checklist-adding-config-fields.md]:
```suggestion
        `agent/config/runtime_test.go`.
```

File: contributing/checklist-adding-config-fields.md [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+        more of the following places:
+         - `ReloadConfig` in `agent/agent.go` if it needs to affect the local
+           client state or another client agent component.
+         - `ReloadConfig` in `agent/consul/client.go` if it needs to affect
+           state for client agent's RPC client.
+      - [ ] Add a test to `agent/agent_test.go` similar to others with prefix
+        `TestAgent_reloadConfig*`.
+ - [ ] **If** the new config field(s) include an array of structs or maps.
+      - [ ] Add the path to the call to `lib.PatchSliceOfMaps` in Parse in
+        `config.go`.
```
Comment (crhino) [inline contributing/checklist-adding-config-fields.md]:
```suggestion
        `agent/config/config.go`.
```

File: contributing/checklist-adding-config-fields.md [hunk 6]
```diff
@@ -0,0 +1,206 @@
...
+      - [ ] If none of the tests in `runtime_test.go` failed before you did that,
```
Comment (crhino) [inline contributing/checklist-adding-config-fields.md]:
```suggestion
      - [ ] If none of the tests in `agent/config/runtime_test.go` failed before you did that,
```

File: contributing/checklist-adding-config-fields.md [hunk 7]
```diff
@@ -0,0 +1,206 @@
...
+
+## Adding a CLI Flag Corresponding to the new Field
+If the config field also needs a CLI flag, then follow these steps.
+
+ - [ ] Do all of the steps in [Adding a Simple Config
+   Field For Client Agents](#adding-a-simple-config-field-for-client-agents).
+ - [ ] Add the new flag to `agent/config/flags.go`.
+ - [ ] Add a test case to TestParseFlags in `agent/config/flag_test.go`.
+ - [ ] Add a test case (or extend one if appropriate) to the table test
+   `TestConfigFlagsAndEdgeCases` in `runtime_test.go` to ensure setting the
```
Comment (crhino) [inline contributing/checklist-adding-config-fields.md]:
```suggestion
   `TestConfigFlagsAndEdgeCases` in `agent/config/runtime_test.go` to ensure setting the
```

**Comment by banks** [comment]
Great feedback @crhino thanks!

> I found that using a gist was not the greatest UX, since it only showed me the raw markdown.

If you name the "file" something that ends in `*.md` then gists will format it as Markdown and include clickable check boxes: https://gist.github.com/banks/35e4af7feb4c8f0db7a8a645de29bde9 (you probably need to fork that to make the checkboxes clickable as it edits the underlying markdown).

Do you think it would help to be more explicit about that in the README?

**Comment by crhino** [comment]
Ah, good to know that gists can do that! I don't think it is necessary to go into that much detail in our README, the essential point is to include the checklist with the PR in my opinion, the workflow used to get there doesn't have to be too detailed.

**Pull request merged**